calico_containers/docker_plugin.py

```diff
@@ -64,85 +65,68 @@ def delete_network():
-            except KeyError:
-                app.logger.info("Couldn't find IPv6 gateway for endpoint %s",
-                                ep_id)
-            else:
-                ep.ipv6_gateway = next_hop6
-        else:
-            app.logger.info("Failed to allocate IPv6 address for endpoint %s",
-                            ep_id)
-
-        ep.profile_id = net_id
-
-        # This iface name must match the code in the Endpoint object.
-        iface = IF_PREFIX + ep_id[:11]
-
-        # Create the veth
-        check_call(['ip', 'link',
-                    'add', ep.name,
-                    'type', 'veth',
-                    'peer', 'name', ep.temp_interface_name()])
+                  #TODO - Enpoint is broken. We don't know
+                  # the interface name (libnetwork decides it) so it
+                  # shouldn't be a mandatory parameter.
+
+    # Create a calico endpoint object which we can populate and return to
+    # libnetwork at the end of this method.
+    ep = Endpoint(ep_id=ep_id,
+                  state="active",
+                  mac=FIXED_MAC,
+                  if_name='cali0')
+    ep.profile_id = net_id
+
+    # This method is split into three phases that have side effects.
+    # 1) Assigning IP addresses
+    # 2) Creating VETHs
+    # 3) Writing the endpoint to the datastore.
+    #
+    # A failure in a later phase attempts to roll back the effects of
+    # the earlier phases.
+
+    # First up is IP assignment. By default we assign both IPv4 and IPv6
+    # addresses.
+    # IPv4 failures may abort the request if the address couldn't be assigned.
+    ipv4_and_gateway(ep)
+    # IPv6 is currently best effort and won't abort the request.
+    ipv6_and_gateway(ep)
+
+    # Next, create the veth.
+    try:
+        create_veth(ep)
+    except CalledProcessError as e:
+        # Failed to create or configure the veth.
+        # Back out the IP assignments and the veth creation.
+        app.logger.exception(e)
+        backout_ip_assignments(ep)
+        remove_veth(ep)
+        abort(500)
 
-        # Set the host end of the veth to 'up' so felix notices it.
-        check_call(['ip', 'link', 'set', iface, 'up'])
+    # Finally, write the endpoint to the datastore.
+    try:
+        client.set_endpoint(hostname, CONTAINER_NAME, ep)
+    except DataStoreError as e:
+        # We've failed to write the endpoint to the datastore.
+        # Back out the IP assignments and the veth creation.
+        app.logger.exception(e)
+        backout_ip_assignments(ep)
+        remove_veth(ep)
+        abort(500)
 
-        # Set the mac as libnetwork doesn't do this for us.
-        check_call(['ip', 'link', 'set',
-                    'dev', ep.temp_interface_name(),
-                    'address', FIXED_MAC])
+    # Everything worked, create the JSON and return it to libnetwork.
+    iface_json = {"ID": 0,
+                  "Address": str(list(ep.ipv4_nets)[0]),
```

Comment:
This line make me nervous about masking bugs that have mutated `ep` in unexpected ways.  Can you add an 

`assert len(ep.ip4_nets) == 1`